plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -735,3 +735,188 @@ func haveSame(a1, a2 []string) bool {
+	//if len(podAffinity.RequiredDuringSchedulingRequiredDuringExecution) != 0 {
+	//	podAffinityTerms = append(podAffinityTerms, podAffinity.RequiredDuringSchedulingRequiredDuringExecution...)
+	//}
+
+	for _, podAffinityTerm := range podAffinityTerms {
+		podAffinityTermMatches, err := checkAnyPodThatMatchPodAffinityTerm(pod, existingPods, node, podAffinityTerm)
+		// if err or podAffinityTerm match failed then return false otherwise continue
+		if err != nil {
+			glog.V(10).Infof("Cannot schedule pod %+v onto node %v, an error ocurred when checking existing pods on the node for PodAffinityTerm %v err: %v",
+				podName(pod), node.Name, podAffinityTerm, err)
+			return false
+		}
+
+		if !podAffinityTermMatches {
+			// Check if it is in special case that the requiredDuringScheduling affinity requirement can be disregarded.
+			// If the requiredDuringScheduling affinity requirement matches a pod's own labels, and there are no other such pods
+			// anywhere, then disregard the requirement.
+			labelSelector, err := unversioned.LabelSelectorAsSelector(podAffinityTerm.LabelSelector)
+			if err != nil || !labelSelector.Matches(labels.Set(pod.Labels)) {
+				glog.V(10).Infof("Cannot schedule pod %+v onto node %v, because not all the existing pods on this node satisfy the PodAffinityTerm %v, err: %+v",
+					podName(pod), node.Name, podAffinityTerm, err)
+				return false
+			}
+
+			// the affinity is to put the pod together with other pods from its same service or RC
+			names := priorityutil.GetNamespacesFromPodAffinityTerm(pod, podAffinityTerm)
+			filteredPods := priorityutil.FilterPodsByNameSpaces(names, allPods)
+			for _, filteredPod := range filteredPods {
+				// if found an existing pod from same service or RC,
+				// the affinity scheduling rules cannot be disregarded.
+				if labelSelector.Matches(labels.Set(filteredPod.Labels)) {
+					glog.V(10).Infof("Cannot schedule pod %+v onto node %v, because not all the existing pods on this node satisfy the PodAffinityTerm %v",
+						podName(pod), node.Name, podAffinityTerm)
+					return false
+				}
+			}
+		}
+	}
+	// all the required pod affinity scheduling rules satisfied
+	glog.V(10).Infof("All the required pod affinity scheduling rules are satisfied for Pod %+v, on node %v", podName(pod), node.Name)
+	return true
+}
+
+// Checks whether the given node has pods which satisfy all the
+// required pod anti-affinity scheduling rules.
+// Also checks whether putting the pod onto the node would break
+// any anti-affinity scheduling rules indicated by existing pods.
+// If node has pods which satisfy all the required pod anti-affinity
+// scheduling rules and scheduling the pod onto the node won't
+// break any existing pods' anti-affinity rules, then return true.
+func checkNodeMatchesHardPodAntiAffinity(pod *api.Pod, existingPods []*api.Pod, node *api.Node, podAntiAffinity *api.PodAntiAffinity) bool {
```

Comment:
This has the same issue as checkNodeMatchesHardPodAffinity(). I think the way to fix it is

1) for the "forward" direction, do the same thing as I in the comment for checkNodeMatchesHardPodAffinity() (this applies to the first call to checkAnyPodThatMatchPodAffinityTerm() from checkNodeMatchesHardPodAntiAffinity())

2) for the "reverse" direction, the loop "for _, ep := range existingPods {" needs to loop over allPods instead of existingPods. I think you do not want to call checkAnyPodThatMatchPodAffinityTerm() from inside that loop. I think the logic you want there is something like
```
for _, p := range allPods {
   // this is pseudocode; obviously you need to consider all of p's anti-affinity terms
   // assume "p.node" means "the node p is running on"
   // assume IfNodeHasTopologyKey() has been rewritten as described in my previous email
   if IfNodeHasTopologyKey(p.node, node, p.podAntiAffinity.podAffinityTerm.TopologyKey) && p.podAntiAffinity.podAffinityTerm.LabelSelector matches pod.Labels { 
       return false // p's anti-affinity would be violated by putting 'pod' on 'node'
    }
}
return true
```